edgedb-tokio/src/tutorial.md

```diff
@@ -0,0 +1,493 @@
+
+Technically arguments can be passed in through any correctly formatted string, but this can be fragile and quickly gets awkward. Note the double `{{` curly braces as well as `' '` around the `movie_title` string to keep the EdgeDB compiler from seeing two separate `movie` and `title` tokens:
+
+```rust
+let movie_title = "Nice movie";
+let release_year = 2023;
+let query = format!(
+    "with movie_title := '{movie_title}',
+year := {release_year},
+movie := (insert Movie {{
+    title := movie_title,
+    release_year := year
+}}),
+select movie {{
+    title,
+    release_year,
+    id
+}}");
+```
+
+Instead, you can pass in arguments as a tuple as the second argument in the client methods. This allows the query to be a directly typed single `&'static str` and the cast notation (`<str>`, `<int32>`) makes the types more clearly visible:
+
+```rust
+let arguments = ("Nice movie", 2023);
+let query = "with 
+movie_title := <str>$0,
+year := <int32>$1,
+movie := (insert Movie {
+  title := movie_title,
+  release_year := year
+})
+  select  {
+    title,
+    release_year,
+    id
+}";
+let query_res: Value = client.query_required_single(query, &(arguments)).await?;
+```
+
+## The `Value` enum
+
+The [Value](https://docs.rs/edgedb-protocol/latest/edgedb_protocol/value/enum.Value.html) enum can be found in the edgedb-protocol crate. A `Value` represents anything returned from EdgeDB so you can always return a `Value` from any of the query methods without needing to deserialize into a Rust type, and the enum can be instructive in getting to know the protocol. On the other hand, returning a `Value` leads to pattern matching to get to the inner value and is not the most ergonomic way to work with results from EdgeDB.
+
+```rust
+pub enum Value {
+    Nothing,
+    Uuid(Uuid),
+    Str(String),
+    Bytes(Vec<u8>),
+    Int16(i16),
+    Int32(i32),
+    Int64(i64),
+    Float32(f32),
+    Float64(f64),
+    BigInt(BigInt),
+    // ... and so on
+}
+```
+
+Most variants of the `Value` enum correspond to a Rust type from the standard library, while some are from the `edgedb-protocol` crate and will have to be constructed. For example, this query expecting a `bigint` will return an error as it receives a `20` (an `i32`) but EdgeDB is expecting a `bigint`:
+
+```rust
+let query = "select <bigint>$0";
+let argument = 20;
+let query_res: Result<Value, _> = client.query_required_single(query, &(argument,)).await;
+assert!(format!("{query_res:?}").contains("expected std::int32"));
+```
+
+Instead, first construct a `BigInt` from the `i32` and pass that in as an argument:
+
+```rust
+use edgedb_protocol::model::BigInt;
+
+let query = "select <bigint>$0";
+let bigint_arg = BigInt::from(20);
+let query_res: Result<Value, _> = client.query_required_single(query, &(bigint_arg,)).await;
+assert_eq!(
+    format!("{query_res:?}"),
+    "Ok(BigInt(BigInt { negative: false, weight: 0, digits: [20] }))"
+);
+```
+
+## Casting inside the EdgeDB compiler
```

Comment:
I'm not sure what this section adds. You should mention that type of the arguments must be provided in the query in the "arguments" section. I don't see a workflow where users would like to read until this section to find out that.

If you think users need some explanation about why type-casting works for variables and for other things, it should belong to EdgeQL docs not Rust docs.